### PR TITLE
Don’t call inactive ticket provider’s class functions

### PR DESCRIPTION
**Pull request by cliffordp**
[ET-843]

_First merge https://github.com/moderntribe/event-tickets/pull/1681 into G20.06 then code review._

7min demo/reasoning: https://share.getcloudapp.com/E0uzPGJP

[ET-843]: https://moderntribe.atlassian.net/browse/ET-843

### CODE REVIEW

File: src/Tribe/Tickets.php [hunk 1]
```diff
@@ -2302,6 +2345,8 @@ public function process_attendee_meta( $product_id, $meta ) {
 		 *
 		 * @internal
 		 *
+		 * @throws ReflectionException
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2348]: :no_entry_sign: Comment missing for @throws tag in function comment

<sub>hash: 53b0628baee50280c310bb7c5a5498b4</sub>

File: src/Tribe/Tickets.php [hunk 2]
```diff
@@ -2921,34 +2966,50 @@ public function ticket_add( $post_id, $data ) {
...
 		/**
-		 * Get the saved or default ticket provider
+		 * Get the saved or default ticket provider.
+		 *
+		 * Will return 'Tribe__Tickets__Tickets' if there is a saved provider that is currently not active.
+		 * Example: If provider is WooCommerce Ticket but ETP is inactive, will return 'Tribe__Tickets__Tickets'.
 		 *
 		 * @since 4.7
+		 * @since TBD If detected module is inactive, return 'Tribe__Tickets__Tickets' to avoid fatals while not
+		 * 				returning unexpected data, such as finding tickets of the wrong provider type.
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2976]: :no_entry_sign: Spaces must be used for mid-line alignment; tabs are not allowed

<sub>hash: acd0bd4f569dbec7d4a93a4b18e181c8</sub>

File: src/Tribe/Tickets.php [hunk 3]
```diff
@@ -2921,34 +2966,50 @@ public function ticket_add( $post_id, $data ) {
...
+			// 'Tribe__Tickets__RSVP' unless filtered but still wouldn't ever be 'Tribe__Tickets__Tickets'.
+			$provider = self::get_default_module();
+
+			// If post ID is set and a value has been saved.
 			if ( ! empty( $event_id ) ) {
 				$saved = get_post_meta( $event_id, $tickets_handler->key_provider_field, true );
 
 				if ( ! empty( $saved ) ) {
-					return $saved;
+					if( tribe_tickets_is_provider_active( $saved ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2994]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after IF keyword; 0 found
* :no_entry_sign: There must be a space between if and the first parenthesis '('

<sub>hash: e9d282d5b7bf07487ab1123e2e897809</sub>

File: src/views/tickets/orders.php [hunk 4]
```diff
@@ -31,7 +31,7 @@
 $post_type                  = get_post_type_object( $event->post_type );
 $user_id                    = get_current_user_id();
 $provider_id                = Tribe__Tickets__Tickets::get_event_ticket_provider( $event_id );
-$provider                   = call_user_func( array( $provider_id, 'get_instance' ) );
+$provider                   = call_user_func( [ $provider_id, 'get_instance' ] );
```
Comment (sc0ttkclark) [inline src/views/tickets/orders.php:34]: Any change in a view requires a `@since` / `@version` update.

File: src/Tribe/Tickets.php [hunk 5]
```diff
@@ -2921,34 +2966,50 @@ public function ticket_add( $post_id, $data ) {
...
+		 * Example: If provider is WooCommerce Ticket but ETP is inactive, will return 'Tribe__Tickets__Tickets'.
 		 *
 		 * @since 4.7
+		 * @since TBD If detected module is inactive, return 'Tribe__Tickets__Tickets' to avoid fatals while not
+		 * 			  returning unexpected data, such as finding tickets of the wrong provider type.
 		 *
 		 * @param int $event_id - the post id of the event the ticket is attached to.
 		 *
-		 * @return string ticket module class name
+		 * @return string|false The ticket module class name (confirmed active). False if module is not active.
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:2980]: Not returning false anymore?

File: src/Tribe/Tickets.php [hunk 6]
```diff
@@ -3204,5 +3263,7 @@ public function edit_global_capacity_level() {
 			_deprecated_function( __METHOD__, '4.6.2', 'wp_send_json_success()' );
 			wp_send_json_success( $data );
 		}
+
+		// @codingStandardsIgnoreEnd
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:3267]: Just adding a note here, the ignore here is wrapped around the deprecated methods, it looks like this PR moves it so it only wraps the deprecated methods as it was in the wrong place before somehow.

File: readme.txt [hunk 7]
```diff
@@ -122,6 +122,7 @@ Currently, the following add-ons are available for Event Tickets:
 * Feature - For the `[tribe-user-event-confirmations]` shortcode, link to both the event and the direct My Tickets page. [ET-792]
 * Fix - Resolve a PHP notice when rendering the Event Tickets Plus field for Birth date [ETP-330]
 * Tweak - Add hook to ticket email template to allow injecting content between ticket details and organizers. [ET-844]
+* Tweak - Add `tribe_tickets_is_provider_active()` so any ticket found in the database can first check its Provider type before calling its functions. For example, if a WooCommerce Ticket was sold in the past but Event Tickets Plus is currently inactive, this would return false. [ET-843]
```
Comment (sc0ttkclark) [inline readme.txt:125]:
```suggestion
* Tweak - Add `tribe_tickets_is_provider_active()` template function so before using any ticket found in the database we can more easily check if the ticket's commerce provider is currently active. [ET-843]
```

File: src/Tribe/Tickets.php (hunk 6 above)
Comment (cliffordp) [reply src/Tribe/Tickets.php:3267]: Right. I'm guessing someone ran PhpStorm reformat code, which I've noticed leaves `//` comments in place if not within a function but moves `/*` comments along with its connected function.

File: src/Tribe/Tickets.php (hunk 5 above)
Comment (cliffordp) [reply src/Tribe/Tickets.php:2980]: No because of always returning at least `'Tribe__Tickets__Tickets'`

File: readme.txt (hunk 7 above)
Comment (cliffordp) [reply readme.txt:125]: implemented in replacement PR

**Pull request closed without merging**